private/optimalNoise.m

```diff
@@ -0,0 +1,123 @@
+%
+% Parameters
+%  iy: index of observable for which the optimal noise parameter
+%  for other inputs see logLikelihoodHierarchical.m and optimalScaling.m
+%
+% Return values:
+%   noise: optimal variance (for Gaussian noise) or scale parameter (for Laplace noise) corresponding to observable i
+%% input assignment
+iy = varargin{1};
+simulation = varargin{2};
+D = varargin{3};
+options = varargin{4};
+s = varargin{5};
+if nargin == 6
+    scale = varargin{6};
+else
+    scale = 'lin';
+end
+
+%% determine scenario
+if(strcmp(options.noise{iy(1)},'multiple') && ...
+        strcmp(options.scaling{iy(1)},'single'))
+    warning(['In options.(',num2str(iy(1)),...
+        ') you combined noise:multiple and scaling:single which is not valid.']);
+end
+
+%% initialization of dimensions
+n_e = size(D,2); %number of experiments
+n_y = size(D(1).my,2); %number of observables
+n_r = size(D(1).my,3); %number of replicates
+if nargout > 1
+    n_theta = size(simulation(1).sy,3); %number of parameters
+end
+
+%% computation of optimal noise
+noise = zeros(1,1,n_r); %variances
+if nargout > 1
+    dnoise = zeros(1,1,n_theta,n_r); %derivatives of the variances
+    if nargout > 2
+        ddnoise = zeros(n_theta,1,n_theta,n_r); %second order derivatives of the variances
+    end
+end
+
+try
+    switch options.noise{iy(1)}
+        case 'multiple'
+            mult_fact_ir = zeros(1,1,n_r);
+            noiseir = zeros(1,1,n_r);
+%             if nargout > 1
+%                 dnoiseir = zeros(1,1,n_theta,n_r);
+%                 if nargout > 2
+%                     ddnoiseir = zeros(n_theta,1,n_theta,n_r);
+%                 end
+%             end
+            for j = 1:n_e %loop over all experiments
+                mult_fact_ir = mult_fact_ir + sum(sum(~isnan(D(j).my(:,iy,:)),1),2);
+                switch scale
+                    case {'log'}
+                        y_sh = bsxfun(@minus,log(D(j).my(:,iy,:)),...
+                            log(bsxfun(@times,s(:,iy,:,j),simulation(j).y(:,iy))));
+                    case {'log10'}
+                        y_sh = bsxfun(@minus,log10(D(j).my(:,iy,:)),...
+                            log10(bsxfun(@times,s(:,iy,:,j),simulation(j).y(:,iy))));
+                    case {'lin'}
+                        y_sh = bsxfun(@minus,D(j).my(:,iy,:),...
+                            bsxfun(@times,s(:,iy,:,j),simulation(j).y(:,iy)));
+                end
+                switch options.distribution
+                    case {'normal'}
+                        %computing the optimal variance 
+                        noiseir = noiseir + sum(nansum(bsxfun(@power,y_sh,2),1),2);
+                    case {'laplace'}
+                        %computing the optimal scale parameter
+                        noiseir = noiseir + sum(nansum(abs(y_sh),1),2);
+                end
+            end
+            
+            noise = bsxfun(@rdivide,noiseir,mult_fact_ir);
+        case 'single'
+            mult_fact_i = zeros(1,1);
```

Comment:
code for `multiple` and `single` are super similar. Should be possible to collapse both cases with a couple of if statements